sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx

```diff
@@ -0,0 +1,129 @@
+# limitations under the License.
+#
+
+# cython: profile=True
+
+""" For internal use only. No backwards compatibility guarantees."""
+
+cimport cython
+from libc.stdint cimport int64_t, INT64_MAX
+from libc.stdlib cimport calloc, free
+
+
+cdef unsigned long long* POWER_TEN = [10e-1, 10e0, 10e1, 10e2, 10e3, 10e4, 10e5,
+                                      10e6, 10e7, 10e8, 10e9, 10e10, 10e11,
+                                      10e12, 10e13, 10e14, 10e15, 10e16, 10e17,
+                                      10e18]
+
+
+cdef int64_t get_log10_round_to_floor(int64_t element):
+  cdef int power = 0
+  while element >= POWER_TEN[power]:
+    power += 1
+  return power - 1
+
+
+cdef class DataflowDistributionCounter(object):
+  """Distribution Counter:
+
+  contains value distribution statistics and methods for incrementing.
+
+  Attributes:
+    min: minimum value of all inputs.
+    max: maximum value of all inputs.
+    count: total count of all inputs.
+    sum: sum of all inputs.
+    first_bucket_offset: starting index of the first stored bucket.
+    last_bucket_offset: end index of buckets.
+    buckets: histogram buckets of value counts for a
+    distribution(1,2,5 bucketing). Max bucket_index is 58( sys.maxint as input).
+    is_cythonized: mark whether DataflowDistributionCounter cythonized.
+  """
+  def __init__(self):
+    self.min = INT64_MAX
+    self.max = 0
+    self.count = 0
+    self.sum = 0
+    self.first_bucket_offset = MAX_BUCKET_SIZE -1
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> calloc(MAX_BUCKET_SIZE, sizeof(int64_t))
+    self.is_cythonized = True
+
+  def __dealloc__(self):
+    """free allocated memory"""
+    free(self.buckets)
+
+  cpdef bint add_input(self, int64_t element) except -1:
+    if element < 0:
+      raise ValueError('Distribution counters support only non-negative value')
+    self.min = min(self.min, element)
+    self.max = max(self.max, element)
+    self.count += 1
+    self.sum += element
+    cdef int64_t bucket_index = self._fast_calculate_bucket_index(element)
+    self.buckets[bucket_index] += 1
+    self.first_bucket_offset = min(self.first_bucket_offset, bucket_index)
+    self.last_bucket_offset = max(self.last_bucket_offset, bucket_index)
+
+  cdef int64_t _fast_calculate_bucket_index(self, int64_t element):
+    """Calculate the bucket index for the given element.
+    
+    Declare calculate_bucket_index as cdef in order to improve performance, 
+    since cpdef will have significant overhead.    
+    """
+    if element == 0:
+      return 0
+    cdef int64_t log10_floor = get_log10_round_to_floor(element)
+    cdef int64_t power_of_ten = POWER_TEN[log10_floor]
+    cdef int64_t bucket_offset = 0
+    if element < power_of_ten * 2:
+      bucket_offset = 0
+    elif element < power_of_ten * 5:
+      bucket_offset = 1
+    else:
+      bucket_offset = 2
+    return 1 + log10_floor * BUCKET_PER_TEN + bucket_offset
```

Comment:
What is the max value that could be returned here. If understand correctly it will be 1 + 18 * 3 + 2 = 57. But `MAX_BUCKET_SIZE` is 59. Not very important, should MAX_BUCKET_SIZE be 58.

Reply:
I would just write MAX_BUCKET_SIZE  = bucket(sys.maxint) here rather than a constant whose value is computed manually.

Reply:
Will we have some downsides if hold a constant value as bucket size?